privacyidea/lib/security/aeshsm.py

```diff
@@ -166,7 +166,18 @@ def random(self, length):
         :param length: length of the random bytestring
         :return:
         """
-        r_integers = self.session.generateRandom(length)
+        retries = 0
+        while True:
+            try:
+                r_integers = self.session.generateRandom(length)
+                break
+            except PyKCS11.PyKCS11Error as exx:
+                log.warning(u"Generate Random failed: {0!s}".format(exx))
+                self.initialize_hsm()
```

Comment:
Hm, it would only call

~~~python
self.pkcs11.lib.C_Initialize()
~~~

which might be enough in some cases. But you are right - we probably would need to re-login.

But you are right, we might need to change

~~~python
def initialize_hsm(self):
    self.is_ready = False
    self.pkcs11.lib.C_Initialize()
    if self.password:
        self._login()
~~~

The thing is, that in case of a password, that is NOT provided in the config file, we need to run ``C_Initialize`` and later (after the python class is created) we need to add the password (and thus ``self._login`` via the ``setup_module`` function.